src/data/error.rs

```diff
@@ -79,3 +81,174 @@ impl<T, E> Recoverable for Result<T, (E, T)> {
         }
     }
 }
+
+#[derive(PartialEq, Debug)]
+pub enum ErrorKind {
+    Warning,
+    Lex,
+    Semantic,
+    Syntax,
+}
+
+pub trait NewCompileError: Any {
+    // Idealy this should be an associated constant, but those are not allowed
+    // in trait objects
+    fn kind(&self) -> ErrorKind;
+
+    // Could also be an associated const
+    fn name(&self) -> &'static str;
+
+    fn message(&self) -> String;
+    fn location(&self) -> Location;
+}
+
+impl dyn NewCompileError {
+    pub fn is<T: NewCompileError>(&self) -> bool {
+        use std::any::TypeId;
+        let t = TypeId::of::<T>();
+        let boxed = self.type_id();
+        t == boxed
+    }
+
+    // TODO: Possibly add a downcast function
+}
+
+impl fmt::Debug for (dyn NewCompileError) {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}: {}", self.name(), self.message())
+    }
+}
+
+impl PartialEq for (dyn NewCompileError) {
+    fn eq(&self, other: &Self) -> bool {
+        self.name() == other.name()
+    }
+}
+
+impl fmt::Display for (dyn NewCompileError) {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{} ({})", self.message(), self.name())
```

Comment:
I would rather this be
```suggestion
        write!(f, "{}: {}", self.name(), self.message())
```
On master, errors currently look like `:1:1: error: invalid program: cannot have empty program`. After your change, they say `:1:1: error: invalid program: cannot have empty program (semantic-error)`. Notice how it says both 'semantic-error' and 'invalid program' - it's telling you the same thing twice.

For the same reason, 'invalid program: ', etc. should be removed in the default errors below.